### ISSUE TITLE
Use Standalone Build Tools when available

### ISSUE THREAD

**Issue by floyd-may** [OmniSharp/omnisharp-roslyn#2114]
Consider having omnisharp use the standalone Visual Studio Build Tools rather than the full Visual Studio. Omnisharp obviously knows that these installs exist, given the logs:

```
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Located 5 MSBuild instance(s)
            1: Visual Studio Build Tools 2017 15.9.28307.1321 - "C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin"
            2: Visual Studio Build Tools 2019 16.9.31105.61 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\MSBuild\Current\Bin"
            3: Visual Studio Community 2019 16.9.31105.61 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin"
            4: Visual Studio Community 2017 15.9.28307.423 - "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin"
            5: StandAlone 16.8.0 - "c:\Users\floyd\.vscode\extensions\ms-dotnettools.csharp-1.23.9\.omnisharp\1.37.6\.msbuild\Current\Bin"
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Registered MSBuild instance: Visual Studio Community 2019 16.9.31105.61 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin"
```

Would it be possible for it to just use the latest of either Visual Studio or the VS Build Tools rather than ignoring the Build Tools installs entirely?

**Comment by filipw** [comment]
it does not "ignore the Build Tools installs entirely". you have the same version in VS installation and in Build Tools.

if the Build Tools had the higher major, minor or patch, it would have been selected. you can see the instance scoring in debug logs

you can also force Omnisharp  to use specific MSBuild using the override section of omnisharp.json.

what is the scenario you are trying to address?

**Comment by floyd-may** [comment]
The logs above are after correcting having a higher version of the build tools installed than of Visual Studio. I had Visual Studio at version 16.8.<something>, and O# selected the StandAlone one that got installed in the extensions folder instead of the VS Build Tools that was at a higher version. So, in my scenario, I had to (undesiredly) install Visual Studio even though I had a higher version of the build tools installed. O# ignored the higher version of the build tools. I'm kicking myself for not keeping the logs prior to the VS install 😖 

They looked something like this:
**FABRICATED LOGS WARNING - NOT REAL LOGS**
```
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Located 5 MSBuild instance(s)
            1: Visual Studio Build Tools 2017 15.9.28307.1321 - "C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin"
            2: Visual Studio Build Tools 2019 16.9.31105.61 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\MSBuild\Current\Bin"
            3: Visual Studio Community 2019 16.<something older> - "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin"
            4: Visual Studio Community 2017 15.9.28307.423 - "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin"
            5: StandAlone 16.8.0 - "c:\Users\floyd\.vscode\extensions\ms-dotnettools.csharp-1.23.9\.omnisharp\1.37.6\.msbuild\Current\Bin"
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Registered MSBuild instance: StandAlone 16.8.0 - "c:\Users\floyd\.vscode\extensions\ms-dotnettools.csharp-1.23.9\.omnisharp\1.37.6\.msbuild\Current\Bin"
```

So, yes, in my scenario, it did, in fact, ignore the Build Tools even though they had a higher version. If anyone else runs into this, please post logs at DEBUG log level.

**Comment by filipw** [comment]
I see. there are additional factors there too like the presence of dotnet core resolvers for example, that's why there is the scoring system so I suspect in your case the VS installation was deemed "more complete"

**Comment by floyd-may** [comment]
@filipw understood, thanks for the detailed answer. I'm sure it feels like you all answer this question repeatedly. Do you know if there's an install option or something with the VS Build Tools that includes the dotnet core resolvers?

Additionally, would you accept a PR that increases the logging verbosity of the MSBuildLocator so that it's more apparent from the logs "this is why O# didn't use this version"?

**Comment by filipw** [comment]
it should be enough to include .NET SDK during the tools installation:

<img width="1157" alt="Bildschirmfoto 2021-03-16 um 16 11 28" src="https://user-images.githubusercontent.com/1710369/111332663-45c71200-8672-11eb-8bc9-fd415d35e322.png">

Yes we could have some extra log messages in the file I linked to that explain why the score ended up what it was - at the moment it will only show you the score results.

**Comment by floyd-may** [comment]
I had originally installed the build tools via chocolatey, so I didn't get a GUI prompt. I just uninstalled Visual Studio 2019 and the VS 2019 build tools, and reinstalled the build tools via chocolatey. This command installs the components that O# is looking for:
```
choco install visualstudio2019buildtools --packageParameters "--add Microsoft.VisualStudio.Workload.NetCoreBuildTools" -y
```

After doing that, O# selects the 2019 build tools:
```
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Located 4 MSBuild instance(s)
            1: Visual Studio Build Tools 2019 16.9.31105.61 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\MSBuild\Current\Bin"
            2: Visual Studio Build Tools 2017 15.9.28307.1321 - "C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin"
            3: Visual Studio Community 2017 15.9.28307.423 - "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin"
            4: StandAlone 16.8.0 - "c:\Users\floyd\.vscode\extensions\ms-dotnettools.csharp-1.23.9\.omnisharp\1.37.6\.msbuild\Current\Bin"
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Registered MSBuild instance: Visual Studio Build Tools 2019 16.9.31105.61 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\MSBuild\Current\Bin"
```

TL; DR: If you don't want to install Visual Studio to support Omnisharp, you can use the above chocolatey install command to install just the build tools that omnisharp needs. Hopefully this helps someone.

**Comment by filipw** [comment]
very nice - thank you
